core/client/src/main/java/org/phenotips/remote/client/internal/RemoteMatchFinder.java

```diff
@@ -78,6 +92,45 @@
     @Inject
     private ConsentManager consentManager;
 
+    @Inject
+    private Provider<XWikiContext> provider;
+
+    private XWikiDocument prefsDoc;
+
+    private List<String> remoteIdsList;
+
+    @Override
+    public void initialize() throws InitializationException
+    {
+        this.remoteIdsList = this.getRemotesList();
+        this.prefsDoc = this.getMatchingRunInfoDoc();
+    }
+
+    private XWikiDocument getMatchingRunInfoDoc()
+    {
+        try {
+            XWikiContext context = this.provider.get();
+            XWikiDocument doc = context.getWiki().getDocument(MATCHING_RUN_INFO_DOCUMENT, context);
+
+            if (doc != null && !doc.isNew()) {
+
+                for (String remoteId : this.remoteIdsList) {
+                    BaseObject object = doc.getXObject(MATCHING_RUN_INFO_CLASS, "serverName", remoteId, false);
+                    if (object == null) {
+                        object = doc.newXObject(MATCHING_RUN_INFO_CLASS, context);
+                        object.setStringValue("serverName", remoteId);
+                    }
+                }
+                context.getWiki().saveDocument(doc, context);
+                return doc;
+            }
```

Comment:
I would add an `else` with `logger.error` to log the problem ("missing matching run info document")